kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala

```diff
@@ -0,0 +1,126 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine
+
+import java.rmi.UnexpectedException
+
+import io.fabric8.kubernetes.api.model.{Pod, PodList}
+import io.fabric8.kubernetes.client.{Config, DefaultKubernetesClient, KubernetesClient, KubernetesClientException}
+import io.fabric8.kubernetes.client.dsl.FilterWatchListDeletable
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf.KUBERNETES_CONTEXT
+import org.apache.kyuubi.engine.ApplicationOperation.{APP_ERROR_KEY, APP_ID_KEY, APP_NAME_KEY, APP_STATE_KEY}
+import org.apache.kyuubi.engine.KubernetesApplicationOperation._
+
+class KubernetesApplicationOperation extends ApplicationOperation with Logging {
+
+  @volatile
+  private var kubernetesClient: KubernetesClient = _
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    info("Start Initialize Kubernetes Client.")
+    val contextOpt = conf.get(KUBERNETES_CONTEXT)
+    if (contextOpt.isEmpty) {
+      warn("Skip Initialize Kubernetes Client, because of Context not set.")
+      return
+    }
+    try {
+      kubernetesClient = new DefaultKubernetesClient(Config.autoConfigure(contextOpt.get))
+      info(s"Initialized Kubernetes Client connect to: ${kubernetesClient.getMasterUrl}")
+    } catch {
+      case e: KubernetesClientException =>
+        error("Fail to init KubernetesClient for KubernetesApplicationOperation", e)
+    }
+  }
+
+  override def isSupported(clusterManager: Option[String]): Boolean = {
+    kubernetesClient != null && clusterManager.nonEmpty &&
+    clusterManager.get.toLowerCase.startsWith("k8s")
+  }
+
+  override def killApplicationByTag(tag: String): KillResponse = {
+    if (kubernetesClient != null) {
+      debug(s"Deleting application info from Kubernetes cluster by $tag tag")
+      try {
+        // Need driver only
+        val operation = findDriverPodByTag(tag)
+        val pod = operation.list().getItems.get(0)
+        (
+          operation.delete(),
+          s"Operation of deleted appId: " +
+            s"${pod.getMetadata.getName} is completed")
+      } catch {
+        case e: Exception =>
+          (false, s"Failed to terminate application with $tag, due to ${e.getMessage}")
+      }
+    } else {
+      throw new IllegalStateException("Methods initialize and isSupported must be called ahead")
+    }
+  }
+
+  override def getApplicationInfoByTag(tag: String): Map[String, String] = {
+    if (kubernetesClient != null) {
+      debug(s"Getting application info from Kubernetes cluster by $tag tag")
+      try {
+        val operation = findDriverPodByTag(tag)
+        val pod = operation.list().getItems.get(0)
+        val res = Map(
+          // Can't get appId, get Pod UID instead.
+          APP_ID_KEY -> pod.getMetadata.getUid,
```

Comment:
We can't get a generic AppId by the usual means. Here we use Kubernetes Pod UID as the identifier